core/ecschema-metadata/src/SchemaFormatsProvider.ts

```diff
@@ -0,0 +1,159 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Metadata
+ */
+
+import { ISchemaLocater, SchemaContext } from "./Context";
+import { SchemaItemKey, SchemaKey } from "./SchemaKey";
+import { SchemaItem } from "./Metadata/SchemaItem";
+import { Format } from "./Metadata/Format";
+import { SchemaItemFormatProps } from "./Deserialization/JsonProps";
+import { BeUiEvent } from "@itwin/core-bentley";
+import { KindOfQuantity } from "./Metadata/KindOfQuantity";
+import { getFormatProps } from "./Metadata/OverrideFormat";
+import { FormatProps, FormatsChangedArgs, FormatsProvider, UnitSystemKey } from "@itwin/core-quantity";
+import { Unit } from "./Metadata/Unit";
+import { InvertedUnit } from "./Metadata/InvertedUnit";
+
+/**
+ * Provides default formats and kind of quantities from a given SchemaContext or SchemaLocater.
+ * @beta
+ */
+export class SchemaFormatsProvider implements FormatsProvider {
+  private _context: SchemaContext;
+  private _unitSystem: UnitSystemKey;
+  private _formatsRetrieved: Set<string> = new Set();
+  public onFormatsChanged = new BeUiEvent<FormatsChangedArgs>();
+  /**
+   *
+   * @param contextOrLocater The SchemaContext or a different ISchemaLocater implementation used to retrieve the schema. The SchemaContext
+   * class implements the ISchemaLocater interface. If the provided locater is not a SchemaContext instance a new SchemaContext will be
+   * created and the locater will be added.
+   */
+  constructor(contextOrLocater: ISchemaLocater, unitSystem: UnitSystemKey) {
+    if (contextOrLocater instanceof SchemaContext) {
+      this._context = contextOrLocater;
+    } else {
+      this._context = new SchemaContext();
+      this._context.addLocater(contextOrLocater);
+    }
+    this._unitSystem = unitSystem;
+  }
+
+  public get context() { return this._context; }
+  public get unitSystem() { return this._unitSystem; }
+
+  public set unitSystem(unitSystem: UnitSystemKey) {
+    this._unitSystem = unitSystem;
+    this.clear();
+  }
+
+  private clear(): void {
+    const formatsChanged = Array.from(this._formatsRetrieved);
+    this._formatsRetrieved.clear();
+    this.onFormatsChanged.raiseEvent({ formatsChanged });
+  }
+
+  private async getKindOfQuantityFormatFromSchema(itemKey: SchemaItemKey, unitSystem: UnitSystemKey): Promise<SchemaItemFormatProps | undefined> {
```

Comment:
I there a point to have `unitSystem` as argument for this function, when it can access `this._unitSystem`?

Reply:
this was a holdover from a previous implementation, I've removed this argument now that we can use `this._unitSystem` directly.